### PR TITLE
Add spread syntax to JsxExpression

### LINKED ISSUE #9495: Add JSX spread children
The JSX specification was just updated (here: https://github.com/facebook/jsx/pull/59) to add for JSX spread children. See the rationale here: https://github.com/facebook/jsx/issues/57

What would be involved in adding support for this new feature of the JSX specification to the TypeScript parser?

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #9495 

This allows you to specify that a JsxExpression should be a list that will be flattened by the JSX consumer.

Thanks @yuit for your 30-minute crash course on how the compiler handles JSX!

### CODE REVIEW

**Comment by sandersn** [comment]
@yuit @RyanCavanaugh can one of you take a look at this?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -10904,7 +10904,11 @@ namespace ts {
 
         function checkJsxExpression(node: JsxExpression) {
             if (node.expression) {
-                return checkExpression(node.expression);
+                const type = checkExpression(node.expression);
+                if (node.dotDotDotToken && !isArrayType(type)) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Check for type being `any`

**Comment by RyanCavanaugh** [comment]
Needs tests for `jsx: react` emit

**Comment by sandersn** [comment]
tsxSpreadChildrenInvalidType.tsx tests `jsx: react` emit (even though it has the incorrect type).
Note that react emit doesn't change because `React.createElement` handles both array and non-array inputs.

**Comment by sandersn** [comment]
@RyanCavanaugh I just noticed that I forgot about this. Is it good to go in?

**Comment by mhegazy** [comment]
@sandersn can you please refresh this PR.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12187,7 +12187,11 @@ namespace ts {
 
         function checkJsxExpression(node: JsxExpression) {
             if (node.expression) {
-                return checkExpression(node.expression);
+                const type = checkExpression(node.expression);
+                if (node.dotDotDotToken && type !== anyType && !isArrayType(type)) {
+                    error(node, Diagnostics.JSX_spread_child_must_be_an_array_type, node.toString(), typeToString(type));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: These arguments are unused :(

**Pull request merged**